In [1]:
# Quiz 4
%load_ext autoreload
%autoreload 2
%matplotlib inline
from my_modules import risk
from my_modules import portfolio
from my_modules import model
from my_modules import data

import numpy as np
import pandas as pd

We will be working with three bonds:

B1 is a 15 Year Bond with a Face Value of $1000 that pays a 5% coupon semi-annually (2 times a year)

B2 is a 5 Year Bond with a Face value of $1000 that pays a 6% coupon quarterly (4 times a year)

B3 is a 10 Year Zero-Coupon Bond with a Face Value of $1000

Assume the yield curve is flat at 5%.

In [2]:
b1 = portfolio.bond_price(12,1000,.05,2,.05)
b2 = portfolio.bond_price(5,1000,.06,4,.05)
b3 = portfolio.bond_price(10,1000,0,1,.05)

1. Which of the three bonds is the most expensive?
 * B2

In [11]:
print ("b1: " + str(b1[0]))
print ("b2: " + str(b2[0]))
print ("b3: " + str(b3[0]))

b1: 1000.0000000000016
b2: 1043.998290335565
b3: 613.9132535407591


2. Which of the three bonds is the least expensive?
 * B3

3. What is the price of the 10 Year Zero Coupon Bond B3?
 * 614

4. Which of the three bonds has the highest (Macaulay) Duration?
n.b. macaulay_duration function gives as output the duration expressed in periods and not in years. If you want to get the yearly duration you need to divide the duration for coupons_per_year;
 * Had difficulties because of not specifying index but guessed the hight/low

In [91]:
b1_cashflow = portfolio.bond_cash_flows(12,1000,.05,2)
b2_cashflow = portfolio.bond_cash_flows(5,1000,.06,4)
b3_cashflow = portfolio.bond_cash_flows(10,1000,0,1)

duration_b1 = portfolio.macaulay_duration(b1_cashflow, 0.05/2)/2
duration_b2 = portfolio.macaulay_duration(b2_cashflow, 0.05/4)/4
duration_b3 = portfolio.macaulay_duration(b3_cashflow, 0.05/1)/1

print("Duration 1: " + str(duration_b1.round(2)))
print("Duration 2: " + str(duration_b2.round(2)))
print("Duration 3: " + str(duration_b3.round(2))) # wrong as longest, should be 1

Duration 1: 9.17
Duration 2: 4.37
Duration 3: 10.0


5. Which of the three bonds has the lowest (Macaulay) Duration?

6. What is the duration of the 5 year bond B2

7. Assume a sequence of 3 liabilities of 100,000, 200,000 and 300,000 that are 3, 5 and 10 years away, respectively. What is the Duration of the liabilities?

In [90]:
liabilities = pd.Series(data=[100_000, 200_000, 300_000], index=[3,5,10])
portfolio.macaulay_duration(liabilities, 0.05)

6.750917852744651

8. Assuming the same set of liabilities as the previous question, build a Duration Matched Portfolio of B1 and B2 to match these liabilities. What is the weight of B2 in the portfolio? (Hint: the code we developed in class erk.match_durations() assumes that all the bonds have the same number of coupons per year. This is not the case here, so you will either need to enhance the code or compute the weight directly e.g. by entering the steps in a Jupyter Notebook Cell or at the Python Command Line)

..skipping

9. Assume you can use any of the bonds B1, B2 and B3 to build a duration matched bond portfolio matched to the liabilities. Which combination of 2 bonds can you NOT use to build a duration matched bond portfolio?
 * B1 and B3 -> over 10 years (unless all b1)

10. Assuming the same liabilities as the previous questions , build a Duration Matched Portfolio of B2 and B3 to match the liabilities.